### PR TITLE
Consolidate handling of "all" for index expression resolver

### PR DESCRIPTION
**Pull request by albertzaharovits**
A refactoring such that callers of
`IndexNameExpressionResolver#resolveExpressions` don't
have to massage parameters when resolving `_all` and
its ilk (empty, null, `*`).

I tried to make it easy to review commit-by-commit.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 1]
```diff
@@ -458,15 +433,27 @@ private void checkSystemIndexAccess(Context context, Set<Index> concreteIndices)
...
+        }
+    }
+
     private static IndexNotFoundException notFoundException(String... indexExpressions) {
-        IndexNotFoundException infe;
-        if (indexExpressions != null && indexExpressions.length == 1) {
-            if (Metadata.ALL.equals(indexExpressions[0])) {
-                infe = new IndexNotFoundException("no indices exist", indexExpressions[0]);
+        final IndexNotFoundException infe;
+        if (indexExpressions == null || indexExpressions.length <= 1) {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:449]: I feel like these clauses could be simplified - eg you're checking for `indexExpressions == null` twice

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 2]
```diff
@@ -1134,7 +1118,11 @@ private WildcardExpressionResolver() {
         public static Collection<String> resolve(Context context, List<String> expressions) {
             Objects.requireNonNull(expressions);
             if (context.getOptions().expandWildcardExpressions() == false) {
-                return expressions;
+                if (expressions.size() == 1 && expressions.iterator().next().equals(Metadata.ALL)) {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:1121]: `expressions` is a List, so you can just do `expressions.get(0)`

**Comment by thecoop** [comment]
Have we got tests covering this behaviour?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 3]
```diff
@@ -320,39 +319,15 @@ String[] concreteIndexNames(Context context, String... indexExpressions) {
...
-        final Set<Index> concreteIndices = Sets.newLinkedHashSetWithExpectedSize(expressions.size());
-        final SortedMap<String, IndexAbstraction> indicesLookup = context.state.metadata().getIndicesLookup();
+        final Set<Index> concreteIndicesResult = Sets.newLinkedHashSetWithExpectedSize(expressions.size());
         for (String expression : expressions) {
-            if (options.ignoreUnavailable() == false) {
+            if (context.getOptions().ignoreUnavailable() == false) {
                 ensureAliasOrIndexExists(context, expression);
             }
-            IndexAbstraction indexAbstraction = indicesLookup.get(expression);
+            final IndexAbstraction indexAbstraction = context.state.metadata().getIndicesLookup().get(expression);
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:330]: What's the reasoning for doing this multi-stage dereference in every loop?

**Comment by albertzaharovits** [comment]
> Have we got tests covering this behaviour?

Which behavior that is?

**Comment by thecoop** [comment]
> > Have we got tests covering this behaviour?
> 
> Which behavior that is?

What this class does - this is a not-entirely-simple refactoring, and it would be good to ensure this has all the same behaviours as pre-refactor.

**Comment by albertzaharovits** [comment]
> What this class does - this is a not-entirely-simple refactoring, and it would be good to ensure this has all the same behaviours as pre-refactor.

It's hard for me to assert that there are tests for everything this class does.

**Comment by thecoop** [comment]
I just spotted `IndexNameExpressionResolverTests`, that seems pretty thorough (not sure how I missed it before). Looks like its all covered then

**Review by grcevski — APPROVED** [review]
LGTM!

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 1 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Good point! pushed [154f450](https://github.com/elastic/elasticsearch/pull/91735/commits/154f45016ed2596574d9ea2c63a020d693305e98)

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 2 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Pushed [71a8a47](https://github.com/elastic/elasticsearch/pull/91735/commits/71a8a4779784780300d212582dbb7ca415a0f7e3)

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 3 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Pushed [c588555](https://github.com/elastic/elasticsearch/pull/91735/commits/c58855555cd8c1a728b2f3a509835662a313d79a) .

**Comment by albertzaharovits** [comment]
Hi @thecoop , thanks for the review. I've addressed your points, please take another look.

**Review by jakelandis — APPROVED** [review]

**Review by thecoop — APPROVED** [review]
Looks good!

**Pull request merged**